Upgrade bootstrap and add example of font-awesome

abhishekbose87 opened the pull request:
Let me know if this looks good.

Issues that I notice
- `react-bootstrap` module does not support bootstrap. Possible alternative is to use [`react-bootstrap-4`](https://www.npmjs.com/package/react-bootstrap-4)
- navbar is not looking that good on mobile

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/shakacode/react-webpack-rails-tutorial/294)

<!-- Reviewable:end -->

justin808 commented:
@jbhatab Please check this out.

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react-webpack-rails-tutorial/294#-:-KK_BmZ97YUqwufabeyz:1533690411)*
<!-- Sent from Reviewable.io -->

justin808 commented:
A couple questions below. Once you have something workable, can you push to your own Heroku instance?

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[client/package.json, line 79 \[r1\]](https://reviewable.io:443/reviews/shakacode/react-webpack-rails-tutorial/294#-KK_CGbqPY94GPBv8Uiv:-KK_CGbqPY94GPBv8Uiw:551002579) ([raw file](https://github.com/shakacode/react-webpack-rails-tutorial/blob/858bf3ca4aa9e9c64eee2d5416b2361d7ab552b3/client/package.json#L79)):*
> ```JSON
>     "react": "^0.14.8",
>     "react-addons-pure-render-mixin": "^0.14.8",
>     "react-bootstrap": "^0.28.5",
> ```

what do we do about react-bootstrap?

---

*[client/webpack.client.base.config.js, line 49 \[r1\]](https://reviewable.io:443/reviews/shakacode/react-webpack-rails-tutorial/294#-KK_Bx5ip5_9JxFYanui:-KK_Bx5ip5_9JxFYanuj:1940785891) ([raw file](https://github.com/shakacode/react-webpack-rails-tutorial/blob/858bf3ca4aa9e9c64eee2d5416b2361d7ab552b3/client/webpack.client.base.config.js#L49)):*
> ```JavaScript
> 
>     new webpack.ProvidePlugin({
>       'window.Tether': 'tether'
> ```

what's this one do?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react-webpack-rails-tutorial/294#-:-KK_CLESSmJ7Osa3ksjI:1257574484)*
<!-- Sent from Reviewable.io -->

abhishekbose87 commented:
I am unsuccesful in pushing the app to Heroku. Can you help me out here ?
I am getting following error message ..

```
remote:        Running: rake assets:precompile
remote:        cd client && npm run build:production
remote:        sh: 1: npm: not found
remote:        rake aborted!
remote:        Command failed with status (127): [cd client && npm run build:production...]
remote:        /tmp/build_dbacc15ff61bc51733b3d85f884b771d/vendor/bundle/ruby/2.3.0/gems/react_on_rails-6.0.2/lib/tasks/assets.rake:91:in `block (3 levels) in <top (required)>'
remote:        Tasks: TOP => assets:precompile => react_on_rails:assets:compile_environment => react_on_rails:assets:webpack
```

I am using tether because of following documentation.
https://github.com/shakacode/bootstrap-loader/blob/40788916f0eb822baa09a5d08beee7e4d9608ebb/README.md#tether

And we can replace `react-bootstrap` with `react-bootstrap-4`.

justin808 commented:
@abhishekbose87 Please try running the precompile locally.

justin808 commented:
@abhishekbose87 Any update? This PR sounds very useful!

jbhatab commented:
@justin808 we removed bootstrap entirely from the gem. We also removed all css based things like font decisions so people can make those decisions on their own. Not sure about merging this one. 

@abhishekbose87 did that compilation error start when you added bootstrap with tether?

justin808 commented on client/.bootstraprc:
```diff
@@ -3,12 +3,16 @@
 
 # Major version of Bootstrap: 3 or 4
-bootstrapVersion: 3
+bootstrapVersion: 4
```
@jbhatab This is the main change. Modernizing to BS4.

justin808 commented:
@jbhatab The point of this tutorial is to show real world integrations of the shakacode open source including:

* https://github.com/shakacode/bootstrap-loader
* https://github.com/shakacode/react_on_rails

jbhatab commented:
wrong repo.... my bad!

Still curious about the compilation issue.

abhishekbose87 commented:
@justin808 Sorry I got busy and conveniently forgot about it 😊  

Let me provide an update by end of this weekend.

abhishekbose87 commented:
@justin808 I tried precompiling the assets and then pushing it to heroku. But it gives the same error.
```
remote:        cd client && npm run build:production
remote:        sh: 1: npm: not found
```

justin808 commented:
@abhishekbose87 did you see: https://github.com/shakacode/react_on_rails/blob/master/docs/additional-reading/heroku-deployment.md

abhishekbose87 commented:
@justin808 Thanks. Got it up and running - https://shrouded-garden-10771.herokuapp.com/

justin808 commented:
@abhishekbose87 Please rebase on top of master. Be sure that you are only moving library versions upwards.

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/shakacode/react-webpack-rails-tutorial/294#-:-KNAZQBIYy5Zc2eHV3zC:1896743549)*
<!-- Sent from Reviewable.io -->

justin808 commented:
@abhishekbose87 Let's get this merged to master! I need you to rebase on top of master, and then I'll merge it!

abhishekbose87 commented:
@justin808 I have rebased master and fixed the conflicts. The new version is deployed to https://shrouded-garden-10771.herokuapp.com/

Let me know if you see any issues.

justin808 commented:
@abhishekbose87 You have linter errors

![2016-08-10_12-41-06](https://cloud.githubusercontent.com/assets/1118459/17573685/c2838c12-5ef7-11e6-8302-8c8b2cea5226.png)

justin808 commented:
@abhishekbose87 Any opinion on if we should wait until react-bootstrap officially supports bootstrap v4?

justin808 commented:
@AlexKVal, what's your opinion of this? This project shows off react-bootstrap...

AlexKVal commented:
The `React-Bootstrap` team's official stance on supporting `Bootstrap@4` [read here](https://github.com/react-bootstrap/react-bootstrap/issues/1187#issuecomment-185338407):

> - we __will__ support bootstrap 4
- we __will not__ start that support until bootstrap 4 is stable enough that we can reasonably develop against it without it constantly changing.

The reason for it is a lot of additional `maintenance burden`.

--
A good example how it usually goes. 😈 

Someone (it is usually _a very kind developer with very good intentions_ 🍒 ) does (a lot of) work
to implement something like that.
(Adding as a dependency a project that is still in `beta` stage.)
https://github.com/react-bootstrap/react-bootstrap/issues/1187#issuecomment-203879424

And a couple of weeks later he/she realizes that although there are **A LOT** of people who would like to __USE__ _the upgraded implementation_,
at the same time there are (usually) __no maintainers__ who is ready to dedicate their time to... __maintain__ it.

And then happens this:
github: `Latest commit d530918  on May 13`
npmjs: `react-bootstrap-4` is published 3 months ago.
The author drops his/her undertaking.

<img width="976" alt="last-commit" src="https://cloud.githubusercontent.com/assets/847572/17629274/32ffe94c-60c2-11e6-90fe-bd619a54009e.png">
<img width="261" alt="npmjs" src="https://cloud.githubusercontent.com/assets/847572/17629275/330368d8-60c2-11e6-88f1-6aaa6915301c.png">

--
If someone from the `react-webpack-rails-tutorial`'s team would merge this upgrade,
then he/she **becomes obliged** to maintain it all the way down 😈 

--
If nobody from the core team of the `react-webpack-rails-tutorial` project is ready
to dedicate their time to maintain `Bootstrap@4-beta` dependency in the project now,
then... it is open-source at the end of the day.

--
Anyone anytime can create fork and maintain `Bootstrap@4` version but everyone
knows that this project is just a tutorial and it is not worth of efforts.

--
Though exactly this PR shows **very good intentions from @abhishekbose87**

🍒

justin808 commented:
Let's keep this on in the PR bucket until bootstrap 4 goes GA! This PR will serve as a reference for those that wish to try out BS4 sooner.

justin808 commented:
Would be great to get this updated! See #238.
